rust/chains/hyperlane-ethereum/src/contracts/multicall.rs

```diff
@@ -34,16 +42,45 @@ pub async fn build_multicall<M: Middleware + 'static>(
     Ok(multicall)
 }
 
-pub fn batch<M: Middleware, D: Detokenize>(
+#[derive(new)]
+pub struct MulticallContractCall<M: Middleware, D: Detokenize> {
+    pub call: ContractCall<M, D>,
+    pub individual_gas_estimates_sum: Option<U256>,
+}
+
+pub async fn batch<M, D>(
     multicall: &mut Multicall<M>,
     calls: Vec<ContractCall<M, D>>,
-) -> ContractCall<M, Vec<MulticallResult>> {
+) -> MulticallContractCall<M, Vec<MulticallResult>>
+where
+    M: Middleware + 'static,
+    D: Detokenize,
+{
     // clear any calls that were in the multicall beforehand
     multicall.clear_calls();
 
-    calls.into_iter().for_each(|call| {
+    let mut individual_estimates_sum = Some(U256::zero());
+    let overhead_per_call: U256 = MULTICALL_OVERHEAD_PER_CALL.into();
+
+    for call in calls.into_iter() {
+        match call.estimate_gas().await {
```

Comment:
do we need to always estimate gas again here? I was hoping we'd be able to tap into the gas estimate we already have for a message from the prepare step

I think we should already have it in the ContractCall from this https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/4fa75a4bd825372ac51fb75ebc66151fe6cdd304/rust/chains/hyperlane-ethereum/src/contracts/mailbox.rs#L418

Maybe we could save on RPCs easily by using `call.tx.gas()` if it's Some, but otherwise falling back to estimating?

Reply:
Also am realizing that when we call `process_contract_call` in `process_batch`, it calls `add_gas_overrides` with the provided Some tx_gas_estimate. This would previously not result in an estimation if tx_gas_estimate was Some, but now that we do the max of the tx_gas_estimate and a new estimate_gas, this isn't the case. So this comment is stale https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/4fa75a4bd825372ac51fb75ebc66151fe6cdd304/rust/chains/hyperlane-ethereum/src/contracts/mailbox.rs#L300, and I think we end up calling estimateGas twice per message when batching, and once now when submitting a non-batch. I think it'd probably be nice to cut down on extra estimateGas calls if we don't need them